pow/src/main/java/tech/pegasys/teku/pow/Eth1Provider.java

```diff
@@ -50,11 +52,25 @@
 
   SafeFuture<Block> getLatestEth1Block();
 
+  SafeFuture<EthBlock.Block> getGuaranteedLatestEth1Block();
+
   SafeFuture<EthCall> ethCall(String from, String to, String data, UInt64 blockNumber);
 
   SafeFuture<BigInteger> getChainId();
 
   SafeFuture<Boolean> ethSyncing();
 
   SafeFuture<List<LogResult<?>>> ethGetLogs(EthFilter ethFilter);
+
+  static String generateEth1ProviderId(final int priority, final String endpoint) {
+    String hostname;
+    try {
+      String tmp = Splitter.on("://").splitToList(endpoint).get(1);
+      hostname = Splitter.on("/").splitToList(tmp).get(0);
+    } catch (Exception e) {
+      hostname = "unknown";
+    }
```

Comment:
We probably shouldn't reinvent URL parsing here.  I'd suggest something like:

```suggestion
    String hostname;
    try {
      final URI uri = new URI(endpoint);
      if (uri.getPort() != - 1) {
        hostname = uri.getHost() + ":" + uri.getPort();
      } else {
        hostname = uri.getHost();
      }
    } catch (URISyntaxException e) {
      hostname = "unknown";
    }
```

Reply:
yeah. don't know why I did that :)